src/prop/sat_proof_manager.h

```diff
@@ -0,0 +1,559 @@
+ *
+ *  (or l3 ~l2)     ~l3
+ *  -------------------- CHAIN_RES
+ *        ~l2
+ *
+ * being added to the LazyCDProofChain. In explaining ~l3 the step S5
+ *
+ *  (or l1 ~l3)      ~l1
+ *  --------------------- CHAIN_RES
+ *         ~l3
+ *
+ * is added. Since ~l1 has no explanation, the process halts. Note that at this
+ * point the step S6 has not been added to the LazyCDProofChain. This is because
+ * to explain ~l5 we need to look at the literal premises in the proofs of
+ * learned clauses.
+ *
+ * The last thing done then in case (3), after the resolution on the conflict
+ * clause and an initial recursive explanation of propagated literals, is to
+ * connect the refutation proof and then recursively its propagated literal
+ * leaves, repeating until a fix point.
+ *
+ * In the above example the first connection yields
+ *
+ *                           (or l1 ~l3 l5)   ~l5
+ *                           ---------------------- CHAIN_RES
+ *                                     (or l1 ~l3)              ~l1
+ *                                     ----------------------------- CHAIN_RES
+ *                      (or l3 ~l2)     ~l3
+ *                      -------------------- CHAIN_RES
+ *  (or l1 l2)   ~l1    ~l2
+ * --------------------------- CHAIN_RES
+ *                     false
+ *
+ * whose literal leaves are ~l1 and ~l5. The former has no explanation, but the
+ * latter is explained by (or l1 ~l5), thus leading to step S6
+ *
+ *  (or l1 ~l5)      ~l1
+ *  --------------------- CHAIN_RES
+ *         ~l5
+ *
+ * then the final connection
+ *
+ *                                             (or l1 ~l5)  ~l1
+ *                                             ---------------- CHAIN_RES
+ *                            (or l1 ~l3 l5)   ~l5
+ *                            ---------------------- CHAIN_RES
+ *                                      (or l1 ~l3)              ~l1
+ *                                      ----------------------------- CHAIN_RES
+ *                       (or l3 ~l2)     ~l3
+ *                       -------------------- CHAIN_RES
+ *   (or l1 l2)   ~l1    ~l2
+ *  --------------------------- CHAIN_RES
+ *                      false
+ *
+ * which has no more explainable literals as leaves.
+ *
+ */
+class SatProofManager
+{
+ public:
+  SatProofManager(Minisat::Solver* solver,
+                  TheoryProxy* proxy,
+                  context::UserContext* userContext,
+                  ProofNodeManager* pnm);
+
+  /** Marks the start of a resolution chain.
+   *
+   * The given clause, at the node level, is registered in d_resLinks with a
+   * null pivot, since this is the first link in the chain.
+   *
+   * @param start a SAT solver clause (list of literals) that will be the first
+   * clause in a resolution chain.
+   */
+  void startResChain(const Minisat::Clause& start);
+  /** Adds a resolution step with a clause
+   *
+   * The resolution step is added to d_resLinks with the clause, at the node
```

Comment:
Do indicate if these methods have some dependency on the other methods (occurring in between startResChain/endResChain).  Its not clear from this whether there is a dependency.